bpf/tests/hairpin_flow.c

```diff
@@ -0,0 +1,318 @@
+	/* Init packet builder */
+	pktgen__init(&builder, ctx);
+
+	/* Push ethernet header */
+	l2 = pktgen__push_ethhdr(&builder);
+
+	if (!l2)
+		return TEST_ERROR;
+
+	ethhdr__set_macs(l2, src, dst);
+
+	/* Push IPv4 header */
+	l3 = pktgen__push_default_iphdr(&builder);
+
+	if (!l3)
+		return TEST_ERROR;
+	l3->saddr = v4_pod_one;
+	l3->daddr = v4_svc_one;
+
+	/* Push TCP header */
+	l4 = pktgen__push_default_tcphdr(&builder);
+
+	if (!l4)
+		return TEST_ERROR;
+	l4->source = tcp_src_one;
+	l4->dest = tcp_svc_one;
+
+	data = pktgen__push_data(&builder, default_data, sizeof(default_data));
+
+	if (!data)
+		return TEST_ERROR;
+
+	/* Calc lengths, set protocol fields and calc checksums */
+	pktgen__finish(&builder);
+
+	/* Register a fake LB backend with endpoint ID 124 for our service */
+	lb_svc_key.address = v4_svc_one;
+	lb_svc_key.dport = tcp_svc_one;
+	lb_svc_key.scope = LB_LOOKUP_SCOPE_EXT;
+
+	/* Create a service with only one backend */
+	lb_svc_value.count = 1;
+	lb_svc_value.flags = SVC_FLAG_ROUTABLE;
+	lb_svc_value.rev_nat_index = revnat_id;
+	map_update_elem(&LB4_SERVICES_MAP_V2, &lb_svc_key, &lb_svc_value, BPF_ANY);
+
+	/* Insert a reverse NAT entry for the above service */
+	revnat_value.address = v4_svc_one;
+	revnat_value.port = tcp_svc_one;
+	map_update_elem(&LB4_REVERSE_NAT_MAP, &revnat_id, &revnat_value, BPF_ANY);
+
+	/* A backend between 1 and .count is chosen, since we have only one backend
+	 * it is always backend_slot 1. Point it to backend_id 124.
+	 */
+	lb_svc_key.backend_slot = 1;
+	lb_svc_value.backend_id = 124;
+	map_update_elem(&LB4_SERVICES_MAP_V2, &lb_svc_key, &lb_svc_value, BPF_ANY);
+
+	/* Create backend id 124 which contains the IP and port to send the
+	 * packet to.
+	 */
+	backend.address = v4_pod_one;
+	backend.port = tcp_svc_one;
+	backend.proto = IPPROTO_TCP;
+	backend.flags = 0;
+	map_update_elem(&LB4_BACKEND_MAP_V2, &lb_svc_value.backend_id, &backend, BPF_ANY);
+
+	/* Add an IPCache entry for pod 1 */
+	cache_key.lpm_key.prefixlen = 32;
+	cache_key.family = ENDPOINT_KEY_IPV4;
+	cache_key.ip4 = v4_pod_one;
+	/* a random sec id for the pod */
+	cache_value.sec_label = 112233;
+	map_update_elem(&IPCACHE_MAP, &cache_key, &cache_value, BPF_ANY);
+
+	ep_key.ip4 = v4_pod_one;
+	ep_key.family = ENDPOINT_KEY_IPV4;
+	map_update_elem(&ENDPOINTS_MAP, &ep_key, &ep_value, BPF_ANY);
+
+	/* Jump into the entrypoint */
+	tail_call_static(ctx, &entry_call_map, 0);
+	/* Fail if we didn't jump */
+	return TEST_ERROR;
+}
+
+CHECK("tc", "hairpin_flow_1_forward_v4")
+int hairpin_flow_forward_check(__maybe_unused const struct __ctx_buff *ctx)
+{
+	void *data;
+	void *data_end;
+	__u32 *status_code;
+	struct iphdr *l3;
+	struct tcphdr *l4;
+
+	test_init();
+
+	data = (void *)(long)ctx->data;
+	data_end = (void *)(long)ctx->data_end;
+
+	if (data + sizeof(__u32) > data_end)
+		test_fatal("status code out of bounds");
+
+	status_code = data;
+
+	if (*status_code != TC_ACT_REDIRECT) {
+		test_log("status code = %lu", *status_code);
```

Comment:
As I mentioned in the previous review unsigned long is 64 bits on linux/x86_64. It will get promoted so not a big deal, but just noting

Reply:
Technically, we have our own interpretation of the placeholder values which are platform independent. https://github.com/cilium/cilium/blob/master/test/bpf_tests/bpf_test.go#L423. I have simply removed this log for now since it didn't add much value anyway.